PLFM-9170 - Export Grid as CSV

nickgros opened the pull request:
Add service to export Grid as CSV

nickgros committed:
PLFM-9170

Add ability to get default internal grid connection

nickgros committed:
PLFM-9170

Add service to export Grid as CSV

nickgros commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/grid/GridDaoImpl.java:
```diff
@@ -267,6 +267,19 @@ public List<GridConnectionInfo> listConnections(String sessionId) {
 				CONNECTION_MAPPER, sessionId);
 	}
 
+    @Override
+    public Optional<GridConnectionInfo> getDefaultInternalConnection(String sessionId) {
+        ValidateArgument.required(sessionId, "sessionId");
+        // This will be the largest replica ID that is within the bounds of our internal ID space
+        try {
+            return Optional.ofNullable(jdbcTemplate.queryForObject(
+                    "SELECT * FROM GRID_CONNECTION WHERE SESSION_ID = ? AND SOURCE = 'INTERNAL' ORDER BY REPLICA_ID DESC LIMIT 1",
+                    CONNECTION_MAPPER, sessionId));
+        } catch (EmptyResultDataAccessException e) {
+            return Optional.empty();
+        }
+    }
```
Note: not checking if ID is in range (`<= max internal ID`). Is `SOURCE = 'INTERNAL'` sufficient?

nickgros commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/grid/DownloadFromGridRequest.json:
```diff
@@ -0,0 +1,33 @@
+		},
+		"includeEtag": {
+			"type": "boolean",
+			"description": "Should the first column contain the row etag?  The default value is 'true'."
```
```suggestion
			"description": "Should the first (or third if ROW_ID/VERSION are included) column contain the row etag?  The default value is 'true'."
```

nickgros commented on lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/grid/DownloadFromGridResult.json:
```diff
@@ -0,0 +1,18 @@
+    },
+    "resultsFileHandleId": {
+      "type": "string",
+      "description": "The resulting file handle ID can be used to download the CSV file created by this job.  The file will contain all of the data requested in the query SQL provided when the job was started."
```
```suggestion
      "description": "The resulting file handle ID can be used to download the CSV file created by this job.  The file will contain all of the data requested in the grid session provided when the job was started."
```

nickgros committed:
PLFM-9170

Add unit tests for worker
Fix a few minor typos/TODOs

nickgros committed:
PLFM-9170

fix tests

john-hill commented on client/synapseJavaClient/src/main/java/org/sagebionetworks/client/AsynchJobType.java:
```diff
@@ -72,7 +73,8 @@ public enum AsynchJobType {
-	CreateGrid("/grid/session", CreateGridResponse.class, RestEndpointType.repo);
+    CreateGrid("/grid/session", CreateGridResponse.class, RestEndpointType.repo),
+    GridCsvDownload("/grid/download/csv", DownloadFromGridResult.class, RestEndpointType.repo);
 	;
```
We will need a new test in ITGridControllerTest to ensure that the endpoints work as expected.

john-hill commented on lib/jdomodels/src/main/resources/schema/AsynchJobStatus-ddl.sql:
```diff
@@ -2,7 +2,7 @@ CREATE TABLE IF NOT EXISTS `ASYNCH_JOB_STATUS` (
   `JOB_STATE` ENUM('PROCESSING','FAILED','COMPLETE') NOT NULL,
-  `JOB_TYPE` ENUM('TABLE_UPDATE_TRANSACTION','UPLOAD_CSV_TO_TABLE_PREVIEW','DOWNLOAD_CSV_FROM_TABLE','QUERY','QUERY_NEXT_PAGE','BULK_FILE_DOWNLOAD','MIGRATION','DOI','ADD_FILES_TO_DOWNLOAD_LIST','STORAGE_REPORT','JSON_SCHEMA_CREATE','VIEW_COLUMN_MODEL_REQUEST','GET_VALIDATION_SCHEMA','QUERY_DOWNLOAD_LIST', 'ADD_TO_DOWNLOAD_LIST', 'DOWNLOAD_LIST_PACKAGE', 'DOWNLOAD_LIST_MANIFEST', 'FILE_HANDLE_ARC ...
+  `JOB_TYPE` ENUM('TABLE_UPDATE_TRANSACTION','UPLOAD_CSV_TO_TABLE_PREVIEW','DOWNLOAD_CSV_FROM_TABLE','QUERY','QUERY_NEXT_PAGE','BULK_FILE_DOWNLOAD','MIGRATION','DOI','ADD_FILES_TO_DOWNLOAD_LIST','STORAGE_REPORT','JSON_SCHEMA_CREATE','VIEW_COLUMN_MODEL_REQUEST','GET_VALIDATION_SCHEMA','QUERY_DOWNLOAD_LIST', 'ADD_TO_DOWNLOAD_LIST', 'DOWNLOAD_LIST_PACKAGE', 'DOWNLOAD_LIST_MANIFEST', 'FILE_HANDLE_ARC ...
   `CANCELING` BIT(1) NOT NULL,
```
we do have schema change so if we want to put this into staging, we would need to alter the table.

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImpl.java:
```diff
@@ -0,0 +1,108 @@
+        gridManager.getGridSession(userInfo, request.getSessionId());
+
+        GridConnectionInfo connectionInfo = gridManager.getDefaultInternalConnection(request.getSessionId())
+                .orElseThrow(() -> new RecoverableMessageException("No internal connection found for session: " + request.getSessionId()));
```
This would be an unexpected state.  We create a new connection when we first create the grid.
What happens if there are outstanding patches?  Maybe throw a recoverable for this case?
Move GridReplicaPatchBuilderManagerImpl.getCurrentClock() to the GridIndexManager, then both your code and the patch builder have access to a method to determine if the replica is up-to-date.

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImpl.java:
```diff
@@ -0,0 +1,108 @@
+
+            // Note that the grid has a maximum of 100,000 rows
+            final long ROWS_PER_PAGE = 1_000L;
+            PaginationIterator<RowView> iterator = new PaginationIterator<>(
```
this logic should be in the view manager (returns an iterator)

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImpl.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Override
+    public DownloadFromGridResult exportGridAsCsv(UserInfo userInfo, DownloadFromGridRequest request, CSVWriterStream stream) {
```
make void for now since it has nothing to share.

john-hill commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImpl.java:
```diff
@@ -0,0 +1,108 @@
+        } catch (IOException e) {
+            throw new RuntimeException("Error writing to CSV stream", e);
+        }
+        DownloadFromGridResult result = new DownloadFromGridResult();
```
will move to the caller

john-hill commented on services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImplTest.java:
```diff
@@ -0,0 +1,248 @@
+
+        rowViewPage = new ArrayList<>();
+        rowViewPage.add(new RowView().setRowObject(new RowObject()
+                .setMetadata(new RowMetadata().setSynapseRow(new SynapseRow().setRowId(1L).setVersionNumber(1L).setEtag("etag1")))
```
duplicate numbers could hide bug.

john-hill commented on services/workers/src/main/java/org/sagebionetworks/grid/workers/GridCSVDownloadWorker.java:
```diff
@@ -0,0 +1,95 @@
+        File temp = null;
+        try {
+            // Since each row must first be read from the database then uploaded to S3
+            long totalProgress = 100;
```
can we move "business logic" to the "manager"

john-hill commented on services/workers/src/test/java/org/sagebionetworks/grid/workers/GridCSVDownloadWorkerTest.java:
```diff
@@ -0,0 +1,155 @@
+        RuntimeException error = new RuntimeException("Bad stuff happened");
+        // table not available
+        when(mockGridReplicaCsvExporter.exportGridAsCsv(any(),any(), any())).thenThrow(error);
+        assertThrows(RuntimeException.class, () -> {
```
runtime should be logged but not thrown (worker framework will just log a generic stack trace).

nickgros committed:
PLFM-9170 - code review changes

- Check that patches have been applied before running CSV
- Move business logic from GridCSVDownloadWorker to GridReplicaCsvExporter
- Add IT test for Grid CSV export
- Refactor to expose `getCurrentClockIfAllPatchesApplied` in `GridIndexManager`
- Refactor to move `CSVWriterProvider` and implementation logic to `lib-utils` so it can be used outside of `services-workers`/`lib-table-cluster`

nickgros commented on lib/lib-grid-db/pom.xml:
```diff
@@ -24,6 +24,11 @@
 		<dependency>
+            <groupId>org.sagebionetworks</groupId>
+            <artifactId>lib-jdomodels</artifactId>
+        </dependency>
```
`GridIndexManagerImpl` in `lib-grid-db` now needs `GridDao` in `lib-jdomodels`

nickgros commented on lib/lib-grid-db/src/test/java/org/sagebionetworks/grid/db/GridIndexManagerImplTest.java:
```diff
@@ -216,4 +220,42 @@ public void testIsPatchAlreadyAppliedWithNewerReplica() {
 		assertTrue(manager.isPatchAlreadyApplied(sessionId, replicaId, patch.getPatchId()));
 	}
 
+    @Test
```
Tests were moved from `gridIndexDao`

nickgros commented on lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/utils/CSVUtils.java:
```diff
@@ -190,43 +190,4 @@ public static String guessContentType(String separator) {
-	 * @param request
-	 * @return
-	 */
-	public static CSVWriter createCSVWriter(Writer writer, CsvTableDescriptor csvTableDescriptor) {
```
Moved logic to csv-utils to be used without depending on `lib-table-cluster`

nickgros commented on lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/utils/CSVUtilsTest.java:
```diff
@@ -439,105 +439,6 @@ public void testCreateCSVReaderAllOverride(){
 	}
-	
-
-	@Test
```
moved

nickgros commented on lib/lib-utils/pom.xml:
```diff
@@ -62,6 +62,11 @@
 			<artifactId>csv-utilities</artifactId>
 		</dependency>
 
+        <dependency>
+            <groupId>org.sagebionetworks</groupId>
+            <artifactId>lib-auto-generated</artifactId>
+        </dependency>
```
Uses `CsvTableDescriptior` which is in `lib-auto-generated`

nickgros commented on lib/lib-utils/src/main/java/org/sagebionetworks/util/csv/CSVWriterProviderImpl.java:
```diff
@@ -0,0 +1,45 @@
+public class CSVWriterProviderImpl implements CSVWriterProvider {
+
+	@Override
+	public CSVWriter createWriter(Writer writer, CsvTableDescriptor csvTableDescriptor) {
```
`CSVWriterProvider` and `Impl` were moved from `services-workers` to `lib-utils`. The previous implementation called a static method in `lib-table-cluster`; this implementation was moved to this file, and is no longer static, to simplify mocking.

nickgros commented on lib/lib-utils/src/test/java/org/sagebionetworks/util/csv/CSVWriterProviderImplTest.java:
```diff
@@ -0,0 +1,117 @@
+import au.com.bytecode.opencsv.CSVWriter;
+import au.com.bytecode.opencsv.Constants;
+
+public class CSVWriterProviderImplTest {
```
The previous implementation just called a static utility, so there were no tests. The static utility implementation was moved to this impl, so the tests for that utility were moved here.

nickgros commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/download/DownloadListManagerImpl.java:
```diff
@@ -685,7 +688,7 @@ String buildManifestCSV(UserInfo user, CsvTableDescriptor descriptor, File tempT
 	 */
 	CSVWriter createCSVWriter(CsvTableDescriptor descriptor, File temp) throws IOException {
-		return CSVUtils.createCSVWriter(
+		return csvWriterProvider.createWriter(
```
Static utility was removed; use the provider bean

nickgros commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/change/GridReplicaPatchBuilderManagerImpl.java:
```diff
@@ -78,14 +75,4 @@ void processChanges(ChangePatchBuilder builder, List<IntendedChange> changes) {
 		}
 	}
 
-	Optional<LogicalTimestamp> getCurrentClock(String sessionId, Long replicaId) {
```
Moved to `GridIndexManager` and renamed `getCurrentClockIfAllPatchesApplied`

nickgros commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImpl.java:
```diff
@@ -1,53 +1,116 @@
+                .orElseThrow(() -> new IllegalStateException("No internal connection found for session: " + gridSessionId));
+
+        gridIndexManager.getCurrentClockIfAllPatchesApplied(connectionInfo.getSessionId(), connectionInfo.getReplicaId())
+                .orElseThrow(() -> new RecoverableMessageException("Current clock could not be retrieved, patches are still being applied to sessionId: " + connectionInfo.getSessionId() + ", replicaId: " + connectionInfo.getReplicaId()));
```
Now check that all patches are applied before querying the grid view

nickgros committed:
PLFM-9170 - fix autowire issue in test

nickgros committed:
PLFM-9170

- fix spring autowire issue
- fix mismatched return type in new SynapseClient method
- treat no grid connection as a recoverable exception

nickgros commented on services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/grid/internal/replica/export/GridReplicaCsvExporterImpl.java:
```diff
@@ -1,53 +1,116 @@
-                .orElseThrow(() -> new RecoverableMessageException("No internal connection found for session: " + request.getSessionId()));
-        GridHeader header = gridReplicaViewManager.readHeader(connectionInfo.getSessionId(), connectionInfo.getReplicaId())
-                .orElseThrow(() -> new RecoverableMessageException("Grid header has not yet been instantiated for sessionId: " + request.getSessionId()));
+        GridConnectionInfo connectionInfo = gridManager.getDefaultInternalConnection(gridSessionId)
+                .orElseThrow(() -> new RecoverableMessageException("No internal connection found for session: " + gridSessionId));
```
We discussed changing this to throw a 500 error, but that caused some of our integration tests to fail when the connection had not been created when the call was made. It seems appropriate to keep this as a `RecoverableMessageException`, but we could instead `sleep` in the tests.

john-hill commented on integration-test/src/test/java/org/sagebionetworks/ITGridControllerTest.java:
```diff
@@ -83,14 +123,87 @@ public void testPingGrid() throws AssertionError, SynapseException, URISyntaxExc
+
+        DownloadFromGridResult result;
+        boolean success = false;
+        while (!success) {
```
UseTimeUtils.waitFor() when waiting in a test.

john-hill commented on lib/lib-grid-db/src/main/java/org/sagebionetworks/grid/db/GridIndexManagerImpl.java:
```diff
@@ -27,11 +28,13 @@ public class GridIndexManagerImpl implements GridIndexManager {
 	private static final Logger log = LogManager.getLogger(GridIndexManagerImpl.class);
 
 	private final GridIndexDao dao;
+    private final GridDao gridDao;
```
We want to export the grid-db project to be an independent project in Github, so any dependency on Synapse needs to be removed.

john-hill commented on lib/lib-grid-db/src/main/resources/grid-db-spb.xml:
```diff
@@ -14,10 +14,13 @@
 	<!-- Import the stack configuration -->
 	<import resource="stack-configuration.spb.xml" />
-	
+    <import resource="classpath:jdomodels-import.xml" />
```
will be removed.

john-hill reviewed (changes requested):
The grid-db dependency change will need to be addressed.

nickgros committed:
code review changes 2

- remove lib-grid-db -> jdo-models dependency
- code style fixes in tests

nickgros committed:
remove invalid reference
